### PR TITLE
Update Seed File to Generate Notification Data and Implement Email Skipping

### PR DESCRIPTION
**Pull request by hvillero**
## Description

This PR updates the seed file to create a substantial amount of notification data for testing and development purposes. Additionally, it implements a mechanism to skip sending email notifications during the seeding process. This is achieved by using a global configuration controlled by an environment variable.

### Key Changes

1. **Seed File Updates**:
   - The `db/seeds.rb` file has been modified to create 1000 notifications for each type (`NewAircraftCreatedNotification`, `LogbookEntryUpdatedNotification`, `NewLogbookEntryCreatedNotification`, `UserSignUpNotification`, and `UserUpdateNotification`).

2. **Skip Email Notifications**:
   - Introduced a global configuration in `config/application.rb` to control email notifications using the `SEND_EMAIL_NOTIFICATIONS` environment variable.
   - Updated notification classes to check the global configuration before sending email notifications.

### Files Changed

- **config/application.rb**: Added `config.send_email_notifications` controlled by `SEND_EMAIL_NOTIFICATIONS` environment variable.
- **config/environments/development.rb**: Set `SEND_EMAIL_NOTIFICATIONS` to `false` to skip email notifications in development.
- **notification classes**: Updated to conditionally skip email notifications based on the global configuration.
- **db/seeds.rb**: Modified to create a large amount of notification data without sending emails.

#### Set Environment Variable

For development and test environments, set the environment variable to `false` to skip email notifications:

```sh
export SEND_EMAIL_NOTIFICATIONS=false
```

#### Development Environment:
- Run rails db:seed

### CODE REVIEW

File: config/environments/development.rb [hunk 1]
```diff
@@ -93,4 +93,6 @@
   # Raise error when a before_action's only/except options reference missing actions
   config.action_controller.raise_on_missing_callback_actions = true
   config.action_mailer.default_url_options = { host: "localhost", port: 3000 }
+  # Add configuration for skipping email notifications based on environment variable
+  config.send_email_notifications = ENV["SEND_EMAIL_NOTIFICATIONS"] == "true"
```
Comment (jonathanloos) [inline config/environments/development.rb:97]: Can I ask why this change was made? The notifications have a default email send constant with is per-notification (preferred). And this application will only be available in development so we'll likely want to send out the notifications regardless.

My recommendation is to remove this config. @pJeyakumar @srb- thoughts?

File: db/seeds.rb [hunk 2]
```diff
@@ -39,5 +44,43 @@
...
+
+    sign_up_notification.deliver_later(UserSignUpNotification.targets)
+
+    update_notification = UserUpdateNotification.with(event_user: user)
+
+    update_notification.deliver_later(UserUpdateNotification.targets)
+  end
+end
+
+Rails.logger.debug("Seeded 10 users, 20 aircrafts, 1000 logbook entries, and 1000 of each notification.")
```
Comment (jonathanloos) [inline db/seeds.rb:86]:
```suggestion
Rails.logger.debug("Seeded#{users.count} users, #{aircrafts.count} aircrafts,#{logbook_entries.count} logbook entries, and#{Notification.count} of each notification.")
```

**Review by jonathanloos — CHANGES_REQUESTED** [review]

File: config/environments/development.rb (hunk 1 above)
Comment (hvillero) [reply config/environments/development.rb:97]: @jonathanloos : I made this change looking at the way to run the seeds and not trigger all those email notifications during the process, maybe a different solution will be better.

Comment (srb-) [reply config/environments/development.rb:97]: @hvillero did you add this because notification generation was slowing down the seeding too much?

File: db/seeds.rb [hunk 3]
```diff
@@ -39,5 +44,43 @@
 
 # Create Aircrafts
 aircrafts.each do |aircraft_data|
-  Aircraft.create!(aircraft_data)
+  aircraft = Aircraft.create!(aircraft_data)
+  1000.times do
```
Comment (srb-) [inline db/seeds.rb:48]: Might be nice for this to be a constant so we can easily ramp up or down the order of magnitude of the seeding.

File: config/environments/development.rb (hunk 1 above)
Comment (hvillero) [reply config/environments/development.rb:97]: @srb-, sorry for the late response; yes, but also lettter_opener was creating all those files per each email notification.

Comment (hvillero) [reply config/environments/development.rb:99]: Yep, I will remove it, it will be just local.

**Comment by hvillero** [comment]
@jonathanloos, @srb-, @pJeyakumar, all the suggestions have been addressed; thx for the review.

File: app/notifications/logbook_entry_updated_notification.rb [hunk 4]
```diff
@@ -3,7 +3,7 @@ class LogbookEntryUpdatedNotification < Noticed::Base
   deliver_by :email, mailer: "LogbookEntryMailer", method: "updated_email", if: :email_notifications?
 
   DEFAULT_SEND_NOTIFICATION = true
-  DEFAULT_SEND_EMAIL_NOTIFICATION = true
+  DEFAULT_SEND_EMAIL_NOTIFICATION = false
```
Comment (srb-) [inline app/notifications/logbook_entry_updated_notification.rb:6]: I think this will break the RSpec tests? How bad is letting the seeding create these email notifications? Does it make it totally unusable?

Comment (pJeyakumar) [reply app/notifications/logbook_entry_updated_notification.rb:6]: @hvillero can confirm that this breaks some of the RSpec tests.
- after adding the `password { "password" }` under line 5 of  `factories/user.rb` (this is to fix the validation requiring a user to have a password)
- running `rspec` will cause the following tests to fail:
  -  logbook_entry_updated_notification_spec.rb:14
  - logbook_entry_updated_notification_spec.rb:18
  - new_aircraft_created_notification_spec.rb:14
  - new_aircraft_created_notification_spec.rb:18
  - new_logbook_entry_created_notification_spec.rb:18
  - new_logbook_entry_created_notification_spec.rb:22 
  - user_sign_up_notification_spec.rb:11 
  - user_sign_up_notification_spec.rb:15
  - user_update_notification_spec.rb:11 
  - user_update_notification_spec.rb:15

Comment (hvillero) [reply app/notifications/logbook_entry_updated_notification.rb:6]: Yep, I will revert this change and run the specs, thx

**Review by pJeyakumar — APPROVED** [review]

File: db/seeds.rb [hunk 5]
```diff
@@ -41,5 +53,43 @@
...
+    notification = NewAircraftCreatedNotification.with(aircraft: aircraft)
+    notification.deliver_later(NewAircraftCreatedNotification.targets)
+  end
+end
+
+# Create Logbook Entries and Notifications
+LOGBOOK_ENTRY_NOTIFICATIONS.times do
+  logbook_entry = LogbookEntry.create!(
+    aircraft: Aircraft.order("RANDOM()").first,
+    pilot_in_command: users.sample,
```
Comment (pJeyakumar) [inline db/seeds.rb:67]: seems that the `users` variable is not defined yet, running `rails db:seed` eventually gives me:

```
NameError: undefined local variable or method `users' for main (NameError)

    pilot_in_command: users.sample,
```

Comment (hvillero) [reply db/seeds.rb:67]: @pJeyakumar , I just fixed this issue.

File: db/seeds.rb [hunk 6]
```diff
@@ -1,9 +1,21 @@
...
+Notification.destroy_all
+
+USERS_COUNT = 10
+AIRCRAFT_NOTIFICATIONS_COUNT = 1000
+LOGBOOK_ENTRY_NOTIFICATIONS = 1000
+USER_NOTIFICATIONS = 100
+
 # Create Users
-10.times do
+users = USERS_COUNT.times do
```
Comment (pJeyakumar) [inline db/seeds.rb:18]:
```suggestion
users = USERS_COUNT.times.map do
```

need to initialize `users` as an `Array` so that we can call `.sample` on it

File: db/seeds.rb [hunk 7]
```diff
@@ -41,5 +53,43 @@
...
+  )
+  new_logbook_notification = NewLogbookEntryCreatedNotification.with(logbook_entry: logbook_entry)
+  new_logbook_notification.deliver_later(NewLogbookEntryCreatedNotification.targets)
+
+  update_logbook_notification = LogbookEntryUpdatedNotification.with(logbook_entry: logbook_entry)
+  update_logbook_notification.deliver_later(LogbookEntryUpdatedNotification.targets)
 end
+
+# Create 1000 UserSignUpNotification and UserUpdateNotification
+users.all.each do |user|
```
Comment (pJeyakumar) [inline db/seeds.rb:83]:
```suggestion
users.each do |user|
```

not needed anymore since `users` will be an `Array`

**Review by pJeyakumar — CHANGES_REQUESTED** [review]
Almost there, just a few more changes to get it all to run properly!

@hvillero a hint for testing, simply set the following constants like so:
```
USERS_COUNT = 2
AIRCRAFT_NOTIFICATIONS_COUNT = 5
LOGBOOK_ENTRY_NOTIFICATIONS = 5
USER_NOTIFICATIONS = 5
```

and test that the seeding works for smaller numbers records (if it does, then we can be certain that it will work for larger sizes as well)

File: db/seeds.rb [hunk 8]
```diff
@@ -1,9 +1,18 @@
...
-
 require "faker"
 
+# Don't send emails during seed process.
+ActionMailer::Base.perform_deliveries = false
+
+# Clear existing data
+ActiveRecord::Base.connection.execute("TRUNCATE logbook_entries, aircrafts, users, notifications RESTART IDENTITY CASCADE")
+
+USERS_COUNT = 3
```
Comment (pJeyakumar) [inline db/seeds.rb:9]:
```
USERS_COUNT = 10
AIRCRAFT_NOTIFICATIONS_COUNT = 1000
LOGBOOK_ENTRY_NOTIFICATIONS = 1000
USER_NOTIFICATIONS = 100
```

shouldn't push the test values to the repo (we'd just need to make that change locally to test and then discard the change)
- but after confirming that the seeding works with smaller numbers, we'd want our constants to be 10, 1000, etc for when we seed the DB in preparation for the upgrade

**Review by pJeyakumar — CHANGES_REQUESTED** [review]

File: db/seeds.rb (hunk 8 above)
Comment (hvillero) [reply db/seeds.rb]: Sure, the change is done, Thx

**Review by pJeyakumar — APPROVED** [review]

**Review by srb- — APPROVED** [review]
:+1:

**Review by ddnguyen93 — APPROVED** [review]

**Review by jonathanloos — APPROVED** [review]
